### DISCUSSION TITLE
Question about UMI information in the output files vdjca and clns

### DISCUSSION THREAD

**Discussion by silvia1234567890** [milaboratory/mixcr#1470] [Q&A] [answered]
Hello,

for the alignment, I ran this code:

```
mixcr align --preset generic-amplicon \
> --library Scophthalmus_maximus-IGH.json \
> --species Scophthalmus_maximus \
> --rna \
> --rigid-left-alignment-boundary \
> --rigid-right-alignment-boundary C \
> -OsaveOriginalReads=true \
> seqs.fasta \
> alignments.vdjca
```
for the clonotype:
`mixcr assemble alignments.vdjca clones.clns`

and to export:
```
mixcr exportAlignments alignments.vdjca alignments.tsv
mixcr exportClones clones.clns clones.tsv
```

In my clonotype table there is no "UMI count" column, and when I want to check the barcode coverage statistics with:
`mixcr exportQc tags clones.clns tags.pdf`
it gives me this message in the terminal:
`No tag refinement report for clones.clns; did you run refineTagsAndSort command?`

My question is: is there any parameter i forgot to put in the mixcr align code to get this information?

Thank you in advance!

**Comment by mizraelson** [comment]
Hi,

The commands you used do not include UMI parsing and processing, which is why that information is not appearing in the output. Could you please share the library structure and the protocol used for your data so I can assist you with the appropriate set of commands.

Sincerely,
Mark

**Comment by silvia1234567890** [comment]
Hi,
I send you as attached file the library that I built with MiXCR.

I pre-processed the sequences with pRESTO (Immcantation tool) and to do the alignment and assembly I used MiXCR with the commands I shared above. 
The structure of my sequences is similar to the ones used in this pRESTO workflow (which is what I followed to pre-process my sequences): https://presto.readthedocs.io/en/stable/workflows/VanderHeiden2017_Workflow.html

I also tried to do the complete analysis with MiXCR using the raw sequences, and I used this code: 

```
mixcr analyze generic-amplicon-with-umi \
> --library Scophthalmus_maximus-IGH.json \
> --species Scophthalmus_maximus \
> --rna \
> --tag-pattern "^(R1:*) \ ^(UMI:N{15})N{10}(R2:*)" \
> --rigid-left-alignment-boundary \
> --floating-right-alignment-boundary C \
> input_R1.fastq \
> input_R2.fastq \
> results
```
For this case (after do the whole analysis with MiXCR), when I want to run this code:
`mixcr exportQc tags clones.clns tags.pdf`
I get this message in the terminal:
```
App version: 4.5.0; built=Fri Sep 22 14:39:05 CEST 2023; rev=cdb24b4fb7; lib=repseqio.v3.0.1
Libraries have different checksums.
```

Could you help me please?

Thank you in advance!

**Reply by mizraelson** [reply]
I do recommend using MiXCR on raw data. Did you run `analyze` and `exportQc` on the same MiXCR version? Can you share the reports from that analyze command.

**Reply by silvia1234567890** [reply]
Hello,

yes, I'm using the last version of MiXCR (4.5.0) through anaconda. I am sending you as an attachment the reports obtained with the `mixcr analyze` command. 
[analyze reports.txt](https://github.com/milaboratory/mixcr/files/13620935/analyze.reports.txt)

One of the questions is if it is possible to get UMI information in the alignments/clones output tables having preprocessed the raw sequences with another tool and then using the `mixcr align` and `mixcr assemble` commands, as well as the info obtained with `mixcr exportQc tags`.
And other problem I have is that when I used the raw sequences with `mixcr analyze` I got the UMI information in the output tables but when I run the command `mixcr exportQc tags` it doesn't work...

Could you help me please? Thank you!

**Reply by mizraelson** [reply, marked as answer]
Hi,
If your data has already been preprocessed and collapsed by UMI using other tools, MiXCR unfortunately cannot provide UMI statistics, as UMIs are no longer present in the data. To obtain these statistics, the data must be processed with MiXCR from the start. Additionally, MiXCR has more advanced barcode processing algorithms compared to other tools, including pRESTO.

Regarding the issue in the assembly step in your reports:

```
Reads used in clonotypes, percent of total: 377943 (34,32%)
Reads dropped due to the lack of a clone sequence, percent of total: 398086 (36,15%)
```

A significant portion of reads (36%) does not cover the CDR3 region. This could be an issue with the identification of one of the CDR3 boundaries. To investigate this further, I recommend manually examining the alignments using the following command:

```
mixcr exportAlignmentsPretty -s 1000 -n 15 input.vdjca
```

This command will skip 1000 records and export the next 15 alignments. You can then see why CDR3 is not clearly identified in some cases.

Regarding the error you encountered:

```
App version: 4.5.0; built=Fri Sep 22 14:39:05 CEST 2023; rev=cdb24b4fb7; lib=repseqio.v3.0.1
Libraries have different checksums.
```

This issue could potentially be due to a file that has the same name as your library but contains slightly different content, which might be present in one of the MiXCR library directories. I suggest conducting the analysis again in a new, clean folder, starting from the reference preparation stage. Use a different name for the library to ensure there are no duplicates. If you can share one of the samples with us, I would be able to attempt reproducing the error and provide assistance in finding a solution.

Sincerely,
Mark

**Reply by silvia1234567890** [reply]
Thank you so much!! 
Just other question: in some sequences, the segments V, D and J are detected but not CDR3 region. 
For example, for the read_Id 1189,
<img width="1449" alt="readId_1189" src="https://github.com/milaboratory/mixcr/assets/131451763/33b1ac48-05f1-4b1c-a4d7-5193e427e0f4">

When I ran the code:
`mixcr exportAlignmentsPretty -s 1000 -n 15 input.vdjca`

 I get this information: 

```
==============================================================================================

>>> Read ids: 1189

>>> Tags:
>>> UMI: GAAATGCAATTAGTG

>>> Target sequences (input sequences):

Sequence0:
Contains features: 

     0 ACAGCACGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAA 79
       GGGGGGGGGGGGGGGGGGGGGGGGFGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGFGGGGGG

    80 CAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGG 159
       GGGGGGEGGGGGGGGGGGGGFFFGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGFGGGFGGGGGFFGGGGGGGCGFG,@E9

   160 TAGCTACTA 168
       CGGG9999B

Sequence1:
Contains features: FR3, FR2_TO_FR4, CDR2, FR2, DRightPSegment, FR3_TO_FR4, DCDR3Part, 
VCDR3Part, CDR2_TO_FR4, VDJunction

     0 TACACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGAATGAGATCTACTGGAGTTTCATACTACAA 79
       FFFFFFFEFEFFFFFE<;AFFA;FF@FFFGFFGFGGGGGGFGGGGCFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGG

    80 AGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGC 159
       GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGGDGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG

   160 CTGAAGACACTGCTGTGTATTACTGTGCCAGATACTGGGAGTACCTTGACTACTGGGGGAAAGGCACCATGGTCACCATC 239
       GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFF7GGGGGGGGGGGGGGGGGGGGGGGGG

   240 ACCTCGGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGATGCCATGTGGTTCTGGAACT 299
       GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGCCCCC

>>> Gene features that can be extracted from this paired-read: 
FR3, FR2_TO_FR4, CDR2, FR2, DRightPSegment, FR3_TO_FR4, DCDR3Part, VCDR3Part, CDR2_TO_FR4, 
VDJunction

>>> Alignments with V gene:

IGHV3-18-F (total score = 1862.0)
Alignment of Sequence1 (score = 1862.0):
    101 TTCACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGAATGAGATCTACTGGAGCTTCATACTACAA 180
        | ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||||||||||
      0 TACACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGAATGAGATCTACTGGAGTTTCATACTACAA 79
        FFFFFFFEFEFFFFFE<;AFFA;FF@FFFGFFGFGGGGGGFGGGGCFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGG

    181 AGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGC 260
        ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
     80 AGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGC 159
        GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGGDGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG

    261 CTGAAGACACTGCTGTGTATTACTGTGCCAGA 292
        ||||||||||||||||||||||||||||||||
    160 CTGAAGACACTGCTGTGTATTACTGTGCCAGA 191
        GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG

>>> Alignments with D gene:

IGHDM1 (total score = 60.0)
Alignment of Sequence1 (score = 60.0):
     14 TACTGGGAGTAC 25
        ||||||||||||
    192 TACTGGGAGTAC 203
        GGGGGGGGGGGG

>>> Alignments with J gene:

IGHJ4 (total score = 354.0)
Alignment of Sequence1 (score = 354.0):
     23 TTGACTACTGGGGGAAAGGCACCATGGTCACCATCACCTCGGGTATG 69
        ||||||||||||||||||||||||||||||||||||||||||  |  
    205 TTGACTACTGGGGGAAAGGCACCATGGTCACCATCACCTCGGCCACC 251
        GGGGGGGFF7GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG

>>> Alignments with C gene:

No hits.

==============================================================================================
```

Any thoughts about why CDR3 region is not covered here for example?

Thank you in advance

**Reply by mizraelson** [reply]
Can you please rerun that command with MiXCR v4.6 :

```
 mixcr exportAlignmentsPretty -s 1000 -n 15 input.vdjca
```

There was a bug in the output that we have fixed.

This should also answer the question about the clones with no CDR3. Most likely its due to the absence (not aligned) end of FR3 region.

**Reply by silvia1234567890** [reply]
Thank you, I just updated the MiXCR version. 
I ran the command above with the last MiXCR version and I got this (for the same read_Id):
```
>>> Read ids: 1189

>>> Tags:
>>> UMI: GAAATGCAATTAGTG


                                                                                             
Quality   77777777777777777777777777777777777777777777777777777777777777777777777777777777   
Target0 0 ACAGCACGTCAGATTCAGCACAAACCATGTTTCCTGTAGCTCTGCTGCTGTTGGCAGCTGGATCATGTGTGAAGGGTGAA 79  Score

                                                                                               
Quality    77777777777777777777777777777777777777777777777777777777777777777777777767772674    
Target0 80 CAGTTGACACAGCCGACCTCTGTGACTGTGCAGCCAGGTCAACGTCTGACCATCACCTGTCAGGTCTCTTATTCTCTTGG 159  Score

                         
Quality     677744446    
Target0 160 TAGCTACTA 168  Score

               R1><FR2                                        FR2><CDR2           CDR2><FR3        
                Y  T  A  W  I  R  Q  P  A  G  K  G  L  E  W  I  G  M  R  S  T  G  V  S  Y  Y  K    
   Quality     77777777777777775567765776777777777777777777767777777777777777777777777777777777    
   Target1   0 TACACAGCTTGGATCAGACAGCCTGCAGGAAAAGGACTGGAGTGGATTGGAATGAGATCTACTGGAGTTTCATACTACAA 79   Score
IGHV3-18-F 101 tTcacagcttggatcagacagcctgcaggaaaaggactggagtggattggaatgagatctactggagCttcatactacaa 180  1862

                                                                                                   
                 D  S  L  K  N  K  F  S  I  D  L  D  T  S  S  K  T  V  T  L  N  G  Q  N  V  Q      
   Quality     77777777777777777777777777777777777777777777777777777777777777777777777777777777    
   Target1  80 AGATTCATTAAAGAACAAGTTCAGTATCGACTTAGACACTTCCAGCAAAACTGTGACTCTAAATGGACAGAATGTGCAGC 159  Score
IGHV3-18-F 181 agattcattaaagaacaagttcagtatcgacttagacacttccagcaaaactgtgactctaaatggacagaatgtgcagc 260  1862

                                                        DP>                                        
                                  FR3><CDR3  V><D      D><DP                                       
               P  E  D  T  A  V  Y  Y  C  A  R  Y  W  E  Y  L  D  Y  W  G  K  G  T  M  V  T  I     
   Quality     77777777777777777777777777777777777777777777777777777747777777777777777777777777    
   Target1 160 CTGAAGACACTGCTGTGTATTACTGTGCCAGATACTGGGAGTACCTTGACTACTGGGGGAAAGGCACCATGGTCACCATC 239  Score
IGHV3-18-F 261 ctgaagacactgctgtgtattactgtgccaga                                                 292  1862
    IGHDM1  14                                 tactgggagtac                                     25   60
     IGHJ4  23                                              ttgactactgggggaaaggcaccatggtcaccatc 57   354

                    FR4>                                                    
             T  S  A  T                                                     
Quality     777777777777777777777777777777777777777777777777777777766666    
Target1 240 ACCTCGGCCACCCCAAAGGGACCAACTGTGTTTCCTCTGATGCCATGTGGTTCTGGAACT 299  Score
  IGHJ4  58 acctcggGTaTG                                                 69   354

```
Is this okay?
Any change in any of the parameters I used with the prior version (build library, full analysis with MiXCR, alignments or assembly)?

**Reply by mizraelson** [reply]
It seems that the automatic algorithm responsible for constructing the reference library failed to identify the FR4 Begin for this J gene, which led to the CDR3 sequence not being recognized. Additionally, there are missing reference points for the IGHV6-22-F and IGHV4-1-F genes. This issue can occur when these genes significantly differ from those in other species. We will manually update the library file that you have created to address this and I'll provide you with a revised version that you can use. 

No changes in the parameters are required after the update to the latest MiXCR version.

In the meantime, were you able to successfully execute the command `mixcr exportQc tags clones.clns tags.pdf`?

**Reply by silvia1234567890** [reply]
Do you need the fasta files with the segments I built the library with? or is it enough with the library?
Sorry, I haven't executed the `mixcr exportQc tags` command yet, I will do it tomorrow (here in a few hours) and let you know if it worked. 
Thanks again for your help

**Reply by mizraelson** [reply]
The library file should be enough as the sequences are embedded in it.

**Reply by silvia1234567890** [reply]
Hello,

the command `mixcr exportQc tags clones.clns tags.pdf` it worked. I got this file:
[tags.pdf](https://github.com/milaboratory/mixcr/files/13685378/tags.pdf)
This plot doesn't look good, right?

**Reply by mizraelson** [reply]
The plot shows a lack of reads to cover the entire UMI diversity present in the sample, which is also seen in the report:

```
Filter report:
  Number of groups: 641794
  Number of groups accepted: 641794 (100%)
  Total records weight: 791408
  Records weight accepted: 791408 (100%)
  Operator #0:
    Effective threshold: 1.0
    Nested thresholds:
      #0: 2
      #1: 1
```

We can't excluded UMIs covered by 1 read because in that case we well lose the majority of the reads. For that reason the  `Effective threshold` here is `1`.

**Reply by silvia1234567890** [reply]
Hello,
 
for the command `mixcr analyze` , is there any parameter to keep the original header of the raw sequences in the align and/or clonotype output tables?

In my case, there are paired-end sequences.. An example of headers of my raw sequences (in .fasta format):

R1: 
```
>M07574:12:000000000-L6GTM:1:1101:10005:1112 1:N:0:19
>M07574:12:000000000-L6GTM:1:2102:19457:17565 1:N:0:19
```
R2: 
```
>M07574:12:000000000-L6GTM:1:1101:10005:1112 2:N:0:19
>M07574:12:000000000-L6GTM:1:2102:19457:17565 2:N:0:19
```

Thank you in advance!

**Reply by mizraelson** [reply]
Yes, to preserve the header, use:

```
-Malign.parameters.saveOriginalReads=true
```

To add columns for alignments upon export, include:

```
--add-step exportAlignments
--prepend-export-alignments-field -descrsR1
--prepend-export-alignments-field -descrsR2
```

This will result in a command like:

```
mixcr analyze generic-amplicon-with-umi \
	--library Scophthalmus_maximus-IGH.json \
	--species Scophthalmus_maximus \
	--rna \
	--tag-pattern "^(R1:*) \ ^(UMI:N{15})N{10}(R2:*)" \
	--rigid-left-alignment-boundary \
	--floating-right-alignment-boundary C \
	-Malign.parameters.saveOriginalReads=true \
	--prepend-export-alignments-field -descrsR1 \
	--prepend-export-alignments-field -descrsR2 \
        --add-step exportAlignments \
	input_R1.fastq \
	input_R2.fastq \
	results
```

**Reply by silvia1234567890** [reply]
Hi,

I got an error message for the code above:
```
No field with name "-descrsR1".
Here are supported fields with similar names:
- -dGenes
- -dFamilies
- -dBestIdentityPercent

To list all available fields run `mixcr exportClones -h`.
```
<img width="730" alt="Captura de pantalla 2023-12-20 a las 12 40 35" src="https://github.com/milaboratory/mixcr/assets/131451763/395f2520-b555-4e1e-801f-143fd4cc639e">

Is it possible to get also the bestVhit, bestDhit and the bestJhit gene segments in the output tables?

I also ran run `mixcr exportClones -h` and I would like to know if it's possible to get the info below (or some) in my output files with `mixcr analyze` and how could I incorporate to my code:

```
-isProductive <gene_feature>
-rankByTag(Molecule/Cell/Sample)
-cellGroup
-tags(Molecule/Cell/Sample)
```

Thank you in advance! 
(and sorry for so many questions)

**Reply by mizraelson** [reply]
Yes, apologies for the confusion. The options `-descrsR1` and `-descrsR2` are available only for alignments, as a clone comprises many alignments and includes numerous reads. However, there is a method to trace clones back to reads.

When you run the following command:

```
mixcr analyze generic-amplicon-with-umi \
    --library Scophthalmus_maximus-IGH.json \
    --species Scophthalmus_maximus \
    --rna \
    --tag-pattern "^(R1:*) \ ^(UMI:N{15})N{10}(R2:*)" \
    --rigid-left-alignment-boundary \
    --floating-right-alignment-boundary C \
    -Malign.parameters.saveOriginalReads=true \
    -Massemble.clnaOutput=true \
    --add-step exportAlignments \
    --prepend-export-alignments-field -descrsR1 \
    --prepend-export-alignments-field -descrsR2 \
    input_R1.fastq \
    input_R2.fastq \
    results
```

You will receive a `results.alignments.tsv` file with `descrsR1` and `descrsR2` columns. Instead of a `results.clns` file, you'll get a `results.clna` file containing information on both alignments and clones.

Furthermore, by using:

```
mixcr exportAlignments \
    -cloneId \
    results.clna \
    results.alignmentsClones.tsv
```

You will obtain `results.alignmentsClones.tsv` which provides details on both the read header from the original FASTQ file and a `cloneId` corresponding to the ID in the clonotype table. Records with `cloneId -1` were not utilized in the clonotype assembly.

Regarding the columns you mentioned:

1) `--append-export-clones-field -isProductive CDR3` will add a column indicating whether the CDR3 region has out-of-frame (OOF) or stop codons.

2) `--append-export-clones-field -rankByTag Molecule` will add a UMI count rank.

3) The `-cellGroup` column is only relevant for single-cell analyses, applicable when the [`mixcr groupClones`](https://mixcr.com/mixcr/reference/mixcr-groupClones/) function is used.

4) `--append-export-clones-field -tags Molecule` will divide the clones by UMI tag, resulting in multiple records for the same clone, each row with a unique UMI sequence.

5) To add `bestVhit`, `bestDhit` and the `bestJhit` columns use:
```
--append-export-clones-field -vHit
--append-export-clones-field -dHit 
--append-export-clones-field -jHit
```

Sincerely,
Mark

**Comment by mizraelson** [comment]
> Do you need the fasta files with the segments I built the library with? or is it enough with the library? Sorry, I haven't executed the `mixcr exportQc tags` command yet, I will do it tomorrow (here in a few hours) and let you know if it worked. Thanks again for your help

After reviewing the library, it would be beneficial if you could share the initial FASTA files. Additionally, could you clarify which portion of the V genes is included? I noticed the use of `VRegion` parameter in buildLibrary, but in the library file, there are sequences that contain L2. If all the sequences are similar, it is advisable to use `--v-gene-feature VExon2`.

**Reply by silvia1234567890** [reply]
Hello,
I am sending you the fasta files with the gene segments I used to build the library.

And to build the library I ran this code:

```
mixcr buildLibrary --debug \
  --v-genes-from-fasta IGHV.fasta \
  --v-gene-feature VRegion \  
  --d-genes-from-fasta IGHD.fasta \
  --j-genes-from-fasta IGHJ.fasta \
  --chain IGH \
  --taxon-id 52904 \
  --species Scophthalmus_maximus \
  Scophthalmus_maximus-IGH.json.gz
```

Thank you in advance!

**Reply by mizraelson** [reply]
Please find the corrected library attached:

[Scophthalmus_maximus-IGH_v2.json.gz](https://github.com/milaboratory/mixcr/files/13747608/Scophthalmus_maximus-IGH_v2.json.gz)

**Reply by silvia1234567890** [reply]
thank you so much! 
Just to understand, what is the difference with the previous library?

**Reply by mizraelson** [reply]
We just changed the `--v-gene-feature` from `VRegion` to `VExon2` and manually set the coordinates for J gene.

**Reply by silvia1234567890** [reply]
last questions (I promise).

just to be sure I'm understanding well the clonotype output table. 

For example, in the picture attached, for the cloneId 0:
<img width="2076" alt="Captura de pantalla 2024-01-04 a las 15 09 31" src="https://github.com/milaboratory/mixcr/assets/131451763/8078c782-59ac-44a3-b98d-5d4f19d0d684">

- "target sequence" is the sequence of this clone
- "readCount" means that 1182 reads contributed to this clone <-- would this be the same as saying "the number of different sequences with the same UMI"?
- "UMI count" means that 1143 UMIs (different UMIs) could have the "target sequence" of this clone?

Am I right?

Thank you in advance!

**Reply by mizraelson** [reply]
Hi,

The target sequence refers to the sequence of the region used for clone assembly, typically the `CDR3` region, unless otherwise specified using the [--assemble-clonotypes-by](https://mixcr.com/mixcr/reference/overview-mixins-list/#clonotype-assembly-mix-in-options) option.

The `readCount` represents the total number of reads contributing to the clone, while the UMI count indicates the number of unique UMI groups that have contributed. The close numbers between reads and UMIs suggest that, on average, each UMI group is represented by approximately one read.

Feel free to reach out if you have more questions :)

Best regards,
Mark

**Reply by silvia1234567890** [reply]
oh, I understand. 

I would like to know if it's possible with `mixcr analyze` or another command to get the information below after pre-processing and before alignment steps. 
I mean, after getting the unique sequences in the preprocessing step, for each sequence I would like to know its abundance. And with pRESTO, for each unique sequence (with its UMI) it gives me two parameters: CONSCOUNT (the number of reads with the same UMI) & DUPCOUNT (count of UMIs that has the same sequence).

Example (attached picture):

<img width="707" alt="Captura de pantalla 2024-01-05 a las 12 01 38" src="https://github.com/milaboratory/mixcr/assets/131451763/267163ff-c4a8-4d7f-acea-3afcd1e08189">

Is it possible to get this information with MiXCR? 

I ask you this because I did the comparison of pRESTO (preprocessing) + MiXCR (align and clonotype) vs MiXCR (whole process, with `mixcr analyze`) and I got better results with `mixcr analyze` (more sequences etc).

Thank you

**Reply by mizraelson** [reply]
The thing is, MiXCR works differently, and that's why it gives better results. Initially, MiXCR aligns all reads to the reference sequence, extracts a UMI sequence for each read, and assigns it to the alignments. Later, during the `refineTagsAndSort` step, the UMI barcode sequences with PCR and sequencing errors are corrected, and automatic filtering filters out the UMI groups that are covered by a low number of reads (in your case, MiXCR most likely didn't filter any UMIs because most of them are only covered by one read, and by default, MiXCR tries to preserve at least 80% of the data). Then, alignments are assembled first into consensuses within each UMI group and later into clones across different UMIs. Such alignment-based UMI consensus assembly is one of the important advantages of the MiXCR pipeline.

Thus, consensuses are only formed from already aligned reads (alignments).

You can follow this process by exporting alignments with `mixcr exportAlignments` from different intermediate files:

- The 'sample.vdca' file contains reads successfully aligned to the reference, with a UMI barcode assigned to each alignment.
- The 'sample.refined.vdca' holds the alignments with corrected and filtered UMIs.
- The 'sample.clna' file (which is only available if you have specified `-Massemble.clnaOutput=true`) can show you which alignments with what UMIs were used in the assembly process of each clone. Don't forget to add `-cloneId` to the `mixcr exportAlignments` command to backtrack final clonotypes to the alignments. The alignments, where `cloneId` equals `-1` were dropped due to various reasons (e.g. low sequencing quality etc.)

Sincerely,
Mark

**Reply by silvia1234567890** [reply]
thank you so much for such detailed explanations!! I now have a much better understanding of how MiXCR works.

Best regards,
Silvia
